File: app/src/main/java/com/smoothradio/radio/feature/player/ui/PlayerFragment.java
package com.smoothradio.radio.feature.player.ui;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;
import androidx.transition.TransitionManager;

import com.google.android.gms.ads.AdRequest;
import com.smoothradio.radio.MainActivity;
import com.smoothradio.radio.R;
import com.smoothradio.radio.core.ui.RadioViewModel;
import com.smoothradio.radio.core.util.PlayerManager;
import com.smoothradio.radio.core.util.Resource;
import com.smoothradio.radio.feature.player.util.TimerSetterHelper;
import com.smoothradio.radio.databinding.FragmentPlayerBinding;
import com.smoothradio.radio.core.model.RadioStation;
import com.smoothradio.radio.service.StreamService;


public class PlayerFragment extends Fragment {
    MainActivity mainActivity;

    public RadioStation currentStation;

    FragmentPlayerBinding binding;

    public String state = "";
    public CoordinatorLayout coordinatorLayout;


    //For starting service
    Intent eventIntent;
    int stationId;
    FragmentActivity fragmentActivity;
    RadioViewModel radioViewModel;

    PlayerManager playerManager;

    public PlayerFragment() {
    }


    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        fragmentActivity = (FragmentActivity) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mainActivity = (MainActivity) getContext();

        radioViewModel = new ViewModelProvider(fragmentActivity).get(RadioViewModel.class);

        playerManager = mainActivity.getPlayerManager();

        registerBroadcasts();

        //for player state ui updates
        eventIntent = new Intent(StreamService.ACTION_EVENT_CHANGE)
                .setPackage(fragmentActivity.getPackageName());
    }

    private void setupObserver() {
        radioViewModel.getStationIdLivedata().observe(getViewLifecycleOwner(), intResource -> {
            Log.d("PlayerFragment","getStationIdLivedata");
            if (intResource.status == Resource.Status.SUCCESS) {
                stationId = intResource.data;

                if (currentStation == null) getLatestStationUsingSavedId();

                binding.ivLargeLogo.setImageResource(currentStation.getLogoResource());
                binding.tvStationNamePlayerFrag.setText(currentStation.getStationName());
            }
        });

        radioViewModel.getSelectedStation().observe(getViewLifecycleOwner(), station -> {
            currentStation = station;
        });
    }

    private void getLatestStationUsingSavedId() {
        currentStation = new RadioStation(stationId,0, "", "", "", "",true,false );

        int position = mainActivity.getAdapter().getPositionOfStation(stationId);

        if (!(mainActivity.getAdapter().listIsEmpty() || position == RecyclerView.NO_POSITION)) {
            currentStation = mainActivity.getAdapter().getStationAtPosition(position);
        }
    }

    private void registerBroadcasts() {
        IntentFilter eventFilter = new IntentFilter();
        eventFilter.addAction(StreamService.ACTION_EVENT_CHANGE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            fragmentActivity.registerReceiver(new EventReceiver(), eventFilter, Context.RECEIVER_NOT_EXPORTED);
            fragmentActivity.registerReceiver(new MetadataReceiver(),
                    new IntentFilter(StreamService.ACTION_METADATA_CHANGE), Context.RECEIVER_NOT_EXPORTED);
        } else {
            fragmentActivity.registerReceiver(new EventReceiver(), eventFilter);
            fragmentActivity.registerReceiver(new MetadataReceiver(), new IntentFilter(StreamService.ACTION_METADATA_CHANGE));
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mainActivity == null) {
            mainActivity = (MainActivity) getContext();
        }
        radioViewModel.reloadStationId();

        //When we request for a ui state form service it doesn't work until the service starts playing. This is unusual/
        //unexpected behavior so we will manually set the state to preparing here for now.
        if (playerManager.getIsShowingAd()) {
            state = StreamService.StreamStates.PREPARING;
            broadcastState(state);
        } else {
            Intent getStateFromServiceIntent = new Intent(StreamService.ACTION_GET_STATE).setPackage(fragmentActivity.getPackageName());
            fragmentActivity.sendBroadcast(getStateFromServiceIntent);//get ui state from service
        }
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentPlayerBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        setUpUI();

        radioViewModel.getStationId();
        setupObserver();

        return root;
    }

    private void setUpUI() {
        //Banner AD
        AdRequest playerFragAdRequest = new AdRequest.Builder().build();
        binding.adView.loadAd(playerFragAdRequest);

        //Adding OnclickListeners
        binding.ivPlayButton.setOnClickListener(new PlayButtonListener());
        binding.ivRefresh.setOnClickListener(new Refresh());
        binding.ivSetTimer.setOnClickListener(new SetTimerOnclickListener());
    }

    /**
     * `MetadataReceiver` is a `BroadcastReceiver` responsible for receiving and displaying metadata
     * information about the currently playing audio stream.  It listens for broadcasts from
     * `StreamService` containing the title of the current stream and updates the UI accordingly.
     * <p>
     * Specifically, it receives an intent with the extra `StreamService.EXTRA_TITLE`, extracts the
     * title string, truncates it to a maximum of 70 characters, and updates a `TextView` with the
     * truncated title.  It also uses a `TransitionManager` to animate the UI update for a smoother
     * user experience.
     */
    public class MetadataReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            String metadata = intent.getStringExtra(StreamService.EXTRA_TITLE);
            if (metadata != null) {
                String truncatedMetadata = metadata.substring(0, Math.min(metadata.length(), 70));
                binding.tvMetadata.setText(truncatedMetadata);
                TransitionManager.beginDelayedTransition(binding.playerFrag);
            }
        }
    }


    public class EventReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            state = intent.getStringExtra(StreamService.EXTRA_STATE);
            updateUI();
        }
    }

    public class Refresh implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            playerManager.refresh();
        }
    }

    public class PlayButtonListener implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            radioViewModel.setSelectedStation(currentStation);
            playerManager.playOrStop();
        }
    }

    class SetTimerOnclickListener implements View.OnClickListener {
        @Override
        public void onClick(View view) {
            new TimerSetterHelper(fragmentActivity,coordinatorLayout);
        }
    }

    void updateUI() {
        TransitionManager.beginDelayedTransition(binding.playerFrag);

        binding.tvProgress.setText(state);
        binding.lottieLoadingAnimation.setVisibility(View.INVISIBLE);
        binding.equalizerAnimation.setVisibility(View.INVISIBLE);
        binding.tvMetadata.setText("");
        binding.ivPlayButton.setImageResource(R.drawable.playerfragplayicon);

        if (StreamService.StreamStates.PREPARING.equals(state)) {
            binding.lottieLoadingAnimation.setVisibility(View.VISIBLE);
            return;
        }

        if (StreamService.StreamStates.PLAYING.equals(state)) {
            binding.equalizerAnimation.setVisibility(View.VISIBLE);
            binding.ivPlayButton.setImageResource(R.drawable.playerfragpauseicon);
            return;
        }

        if (StreamService.StreamStates.BUFFERING.equals(state)) {
            binding.lottieLoadingAnimation.setVisibility(View.VISIBLE);
            return;
        }

        if (StreamService.StreamStates.ENDED.equals(state) ||
                StreamService.StreamStates.IDLE.equals(state)) {
            return;
        }

        // If unknown state
        binding.tvProgress.setText("");
        binding.tvMetadata.setText("");
    }

    private void broadcastState(String state) {
        eventIntent.putExtra(StreamService.EXTRA_STATE, state);
        fragmentActivity.sendBroadcast(eventIntent);
    }
}

File: app/src/main/java/com/smoothradio/radio/feature/player/ui/PlayerFragment.kt
package com.smoothradio.radio.feature.player.ui

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.os.Bundle
import android.transition.TransitionManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.isInvisible
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.ads.AdRequest
import com.smoothradio.radio.MainActivity
import com.smoothradio.radio.R
import com.smoothradio.radio.core.model.RadioStation
import com.smoothradio.radio.core.ui.RadioViewModel
import com.smoothradio.radio.core.util.PlayerManager
import com.smoothradio.radio.databinding.FragmentPlayerBinding
import com.smoothradio.radio.feature.player.util.TimerSetterHelper
import com.smoothradio.radio.service.StreamService
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class PlayerFragment : Fragment() {

    private lateinit var mainActivity: MainActivity
    private var currentStation: RadioStation? = null
    private var _binding: FragmentPlayerBinding? = null
    private val binding get() = _binding!!

    private var state: String = ""
    private lateinit var coordinatorLayout: CoordinatorLayout
    private lateinit var eventIntent: Intent

    private lateinit var fragmentActivity: FragmentActivity
    private lateinit var radioViewModel: RadioViewModel
    private lateinit var playerManager: PlayerManager

    override fun onAttach(context: Context) {
        super.onAttach(context)
        fragmentActivity = context as FragmentActivity
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mainActivity = requireActivity() as MainActivity
        radioViewModel = ViewModelProvider(fragmentActivity)[RadioViewModel::class.java]
        playerManager = mainActivity.playerManager

        registerBroadcasts()

        eventIntent =
            Intent(StreamService.ACTION_EVENT_CHANGE).setPackage(fragmentActivity.packageName)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentPlayerBinding.inflate(inflater, container, false)
        setUpUI()
        collectFlows()
        return binding.root
    }

    private fun collectFlows() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                radioViewModel.playingStation.collect { radioStation ->
                    currentStation = radioStation ?: getDefaultStation()

                    currentStation?.let { station ->
                        binding.apply {
                            ivLargeLogo.setImageResource(station.logoResource)
                            tvStationNamePlayerFrag.text = station.stationName
                        }
                    }
                }
            }
        }
    }

    private fun getDefaultStation(): RadioStation {
        val stationId = 0
        val position = mainActivity.radioListRecyclerViewAdapter.getPositionOfStation(stationId)
        if (!(mainActivity.radioListRecyclerViewAdapter.listIsEmpty() || position == RecyclerView.NO_POSITION)) {
                val station =mainActivity.radioListRecyclerViewAdapter.getStationAtPosition(position)
            return station
        }
        return RadioStation(0, 0, "", "", "", "", true,false)
    }

    private fun registerBroadcasts() {
        val eventFilter = IntentFilter().apply {
            addAction(StreamService.ACTION_EVENT_CHANGE)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            fragmentActivity.registerReceiver(
                EventReceiver(),
                eventFilter,
                Context.RECEIVER_NOT_EXPORTED
            )
            fragmentActivity.registerReceiver(
                MetadataReceiver(),
                IntentFilter(StreamService.ACTION_METADATA_CHANGE),
                Context.RECEIVER_NOT_EXPORTED
            )
        } else {
            fragmentActivity.registerReceiver(EventReceiver(), eventFilter)
            fragmentActivity.registerReceiver(
                MetadataReceiver(),
                IntentFilter(StreamService.ACTION_METADATA_CHANGE)
            )
        }
    }

    override fun onResume() {
        super.onResume()
        mainActivity = mainActivity ?: (requireActivity() as MainActivity)


        if (playerManager.isShowingAd) {
            state = StreamService.StreamStates.PREPARING
            broadcastState(state)
        } else {
            val getStateFromServiceIntent =
                Intent(StreamService.ACTION_GET_STATE).setPackage(fragmentActivity.packageName)
            fragmentActivity.sendBroadcast(getStateFromServiceIntent)
        }
    }

    private fun setUpUI() = with(binding) {
        adView.loadAd(AdRequest.Builder().build())

        ivPlayButton.setOnClickListener(PlayButtonListener())
        ivRefresh.setOnClickListener(Refresh())
        ivSetTimer.setOnClickListener(SetTimerOnclickListener())
    }

    /**
     * `MetadataReceiver` is a `BroadcastReceiver` responsible for receiving and displaying metadata
     * information about the currently playing audio stream.  It listens for broadcasts from
     * `StreamService` containing the title of the current stream and updates the UI accordingly.
     * <p>
     * Specifically, it receives an intent with the extra `StreamService.EXTRA_TITLE`, extracts the
     * title string, truncates it to a maximum of 70 characters, and updates a `TextView` with the
     * truncated title.  It also uses a `TransitionManager` to animate the UI update for a smoother
     * user experience.
     */
    inner class MetadataReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val metadata = intent?.getStringExtra(StreamService.EXTRA_TITLE)
            metadata?.let {
                val truncatedMetadata = it.substring(0, minOf(it.length, 70))
                binding.tvMetadata.text = truncatedMetadata
                TransitionManager.beginDelayedTransition(binding.playerFrag)
            }
        }
    }

    inner class EventReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            state = intent?.getStringExtra(StreamService.EXTRA_STATE) ?: ""
            updateUI()
        }
    }

    inner class Refresh : View.OnClickListener {
        override fun onClick(view: View?) {
            playerManager.refresh()
        }
    }

    inner class PlayButtonListener : View.OnClickListener {
        override fun onClick(view: View?) {
            val station = currentStation ?: return

            radioViewModel.setSelectedStation(station)
        }
    }

    inner class SetTimerOnclickListener : View.OnClickListener {
        override fun onClick(view: View?) {
            TimerSetterHelper(fragmentActivity, coordinatorLayout)
        }
    }

    private fun updateUI() = with(binding) {
        TransitionManager.beginDelayedTransition(playerFrag)

        tvProgress.text = state
        lottieLoadingAnimation.isInvisible = true
        equalizerAnimation.isInvisible = true
        tvMetadata.text = ""
        ivPlayButton.setImageResource(R.drawable.playerfragplayicon)

        when (state) {
            StreamService.StreamStates.PREPARING,
            StreamService.StreamStates.BUFFERING -> {
                lottieLoadingAnimation.isVisible = true
            }

            StreamService.StreamStates.PLAYING -> {
                equalizerAnimation.isVisible = true
                ivPlayButton.setImageResource(R.drawable.playerfragpauseicon)
            }

            StreamService.StreamStates.ENDED,
            StreamService.StreamStates.IDLE -> { /* no-op */
            }

            else -> {
                tvProgress.text = ""
                tvMetadata.text = ""
            }
        }
    }


    private fun broadcastState(state: String) {
        eventIntent.putExtra(StreamService.EXTRA_STATE, state)
        fragmentActivity.sendBroadcast(eventIntent)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
